Add draft dev team proposal (#1)
srobo/dev-team-proposal: README.md

# Student Robotics Dev Team Proposal

This document aims to solve the important issue of development by outlining the formation of a number of smaller sub-groups within Student Robotics:

> - The Development Coordination Team.

Reviewer: I'm not sure I understand the need for this second-level abstraction? What are their responsibilities that aren't already the scope of the core team? All this does is add an additional level of abstraction (and communications overhead) for  whenever the core team may want to know the status of a project.

Author: The Core Team exists to facilitate the competition. It is outside of their scope, and to be honest too much work for them to also have to manage development for future years. This ensures that development doesn't pause between the end of the competition and the start of the next cycle. It also means that we won't end up with a situation like last year when there were volunteers wanting to develop and being told that they can't because the Core Team has other things to do that are more important.

Reviewer: I suspect you're right. If not only because it allows compartmentalization in the fact that there will likely be people who want to be on 1 team and not the other. 

If volunteers were told development couldn't happen, that was wrong. Development however was highly unlikely to ship. I think the dev team ensures those stay distinct.

> - A number of smaller _project groups_.

Reviewer: There's nothing to say a project group has to be smaller than the number of people on the dev team.
```suggestion
- A number of _project groups_.
```

Author: I agree. If nobody objects, we should commit this change.

Reviewer: It's your proposal, make the change.

Reviewer 2: ![](https://i.imgflip.com/2upjp5.jpg)

## The Development Coordination team

> The Development Coordination team (a.k.a. the _Dev Team_) is a group of volunteers that are chosen annually in a similar fashion to the formation of the _Core Team_, although they may be responsible for projects beyond the annual timescale.

Reviewer: Related to a point above, if the development team also rotates annually, why does that not impact development in the same way that rotating a core team does?

Reviewer 2: No, because the Core Team's rotation is not the cause of its inability to develop long-term projects. It's the Core Team's **responsibilities** being limited to an annual cycle. A Core Team working in the defined interests of the Core Team will in theory not be able to pursue any projects with a timescale greater than one year.

The Dev Team would rotate annually much like the Core Team to allow the easy flow of volunteers, but its **responsibility** would be greater than one year at a time. As such it is both able and encouraged to undertake long term projects.

Author: Project groups do not rotate annually.

> The goals of the _Dev Team_ are to ensure the long-term longevity of Student Robotics by organising projects that enhance the experience for SR’s competitors and volunteers in line with its values, and enable SR’s long term growth. Additionally they will aim to fulfill the values of Student Robotics, in particular our goal of reflecting reality by incorporating the latest technology whenever feasible.

Reviewer 3: > reflecting reality by incorporating the latest technology whenever feasible.

I disagree that "reflecting reality" leads to "incorporating the latest technology". While incorporating newer technologies has benefits, I would propose that there are much stronger engineering principles which are more reflective of reality and may run contrary to this. The main example I would cite is using the right tool for any given job, which may include fitting in well with an existing environment.

This does _not_ mean that we should constrain the Dev Team to legacy systems, however it does mean that they should consider how any new system will interact with those systems.

Reviewer 3: Separately, there is also an argument that existing systems are the environment that most engineers operate in, so always presenting competitors with the latest bleeding-edge stuff may actually move away from reality. Clearly there is a balance to be struck between following the wider trends of industry and putting us in a situation where the Dev Team is incentivised to rebuild our stack ever few years by chasing the latest fad.

Reviewer: I agree that using the latest technology rarely reflects reality. In my experience (both professionally and not), large projects will tend towards the LTS model of technologies. Intentionally _not_ using the latest versions of technology, and instead using something which is slightly older, with the added benefit of stability and support. 

Given a recent example, the latest _feasible_ version of the Linux kernel is 5.0, but I suspect its market share is drastically less than 0.1%.  

The entire final sentence actually duplicates part of the former, which already talks about the values of SR. It seems it can be removed without losing any information about the role of the dev team to SR.

```suggestion
The goals of the _Dev Team_ are to ensure the long-term longevity of Student Robotics by organising projects that enhance the experience for SR’s competitors and volunteers in line with its values, and enable SR’s long term growth.
```

Author: ```suggestion
The goals of the _Dev Team_ are to ensure the long-term longevity of Student Robotics by organising projects that enhance the experience for SR’s competitors and volunteers in line with its values, and enable SR’s long term growth. Additionally they will aim to fulfill the values of Student Robotics, in particular our goal of reflecting reality by incorporating the latest stable technology whenever reasonable. 
```

Reviewer 4: I don't think we should have the word "latest", maybe "industry standard"?

If anything, we're doing better at reflecting reality by having horrible out-dated systems which are too expensive to replace, as that's what really happens in software engineering :) (just joking)

Author: > I don't think we should have the word "latest", maybe "industry standard"?
> 
> If anything, we're doing better at reflecting reality by having horrible out-dated systems which are too expensive to replace, as that's what really happens in software engineering :) (just joking)

One could argue that COBOL is an industry standard. I strongly disagree.

> The goals of the _Dev Team_ are to ensure the long-term longevity of Student Robotics by working on projects that enhance the experience for SR’s competitors and volunteers in line with its values, and enable SR’s long term growth. Additionally they will aim to fulfill the values of Student Robotics, in particular our goal of reflecting reality by incorporating the latest technology whenever feasible.

Reviewer: > The goals of the _Dev Team_ are to ensure the long-term longevity of Student Robotics by working on projects

This reads to me like the dev team will be doing the work themselves rather than orchestrating the project groups?

Reviewer: > in particular our goal of reflecting reality by incorporating the latest technology whenever feasible

I don't think this addition is necessary. Yes it's good to be up to date, but I don't think that should be a specific requirement of the dev team. The dev team should focus on all values of SR equally

Reviewer 2: > The dev team should focus on all values of SR equally

I disagree; the nature of separate teams is to have a specific focus. If all the teams and groups worked on all of the values equally, then the teams would be homogeneous and wouldn't actually do different things.

It makes sense for different teams to focus more on different values, so long as across the organisation as a whole, those values are all met. The core team is not well suited to focus on the value of maintaining up to date technology, as that largely requires projects which fall outside its timescale - it is for this very reason that the dev team was devised. It therefore seems to make sense for the dev team to make up the deficiencies we have seen in the core team, rather than simply focusing on all values equally.

Reviewer 5: > > The goals of the _Dev Team_ are to ensure the long-term longevity of Student Robotics by working on projects

> This reads to me like the dev team will be doing the work themselves rather than orchestrating the project groups?

```suggestion
The goals of the _Dev Team_ are to ensure the long-term longevity of Student Robotics by organising projects that enhance the experience for SR’s competitors and volunteers in line with its values, and enable SR’s long term growth. Additionally they will aim to fulfill the values of Student Robotics, in particular our goal of reflecting reality by incorporating the latest technology whenever feasible.
```

Reviewer 5: > reflecting reality by incorporating the latest technology whenever feasible

This sounds rather strict - does this mean we have to switch to a new web framework twice a month? I would suggest something like more recent/modern technologies.

Author: I agree.

```suggestion
The goals of the _Dev Team_ are to ensure the long-term longevity of Student Robotics by working on projects that enhance the experience for SR’s competitors and volunteers in line with its values, and enable SR’s long term growth. Additionally they will aim to fulfill the values of Student Robotics, in particular our goal of reflecting reality by incorporating the recent technology whenever feasible.
```

Reviewer: > by incorporating the recent technology whenever feasible

This is no longer grammatically correct.

## Project Groups

> A _project group_ is a smaller group of self-selected volunteers who have a particular interest in working on something together. The _project group_ should report progress regularly to the _Dev Team_. The _Dev Team_ oversees the various _project groups_ to ensure that they are on track and aligned with the goals of the charity. _Project groups_ are responsible for maintaining projects that they have.

Reviewer: > smaller group of self-selected volunteers

We should definitely ensure a single person can form a 'project group', either because they have a sole interest in something, or because the project is so short-lived it doens't make sense in splitting it up.

Reviewer 2: I believe this is given in one of the examples, but I agree that we should permit it explicitly.

Author: ```suggestion
A _project group_ is a smaller group of at least one self-selected volunteers who have a particular interest in working on something together. The _project group_ should report progress regularly to the _Dev Team_. The _Dev Team_ oversees the various _project groups_ to ensure that they are on track and aligned with the goals of the charity. _Project groups_ are responsible for maintaining projects that they have.
```

Reviewer 5: Alternative that doesn't break the surrounding grammar:

```suggestion
A _project group_ is a smaller group of one or more self-selected volunteers who have a particular interest in working on something together. The _project group_ should report progress regularly to the _Dev Team_. The _Dev Team_ oversees the various _project groups_ to ensure that they are on track and aligned with the goals of the charity. _Project groups_ are responsible for maintaining projects that they have.
```

### Formation of a Project Group

A group of volunteers decides that they would like to work on a project. They communicate their ideas to the _Dev Team_ and the _Dev Team_ will decide whether the project is beneficial to the organisation. A schedule for the project will be discussed between the _Dev Team_ and the _project group_. A _project group_ may apply to the _Dev Team_ for budget for specific purchases for their project, given sufficient justification. The _Dev Team_ should also keep in mind the future needs of _project groups_ when planning their budget.

Alternatively, the _Dev Team_ may identify the need for a project within the organisation and will find interested volunteers to form a group to do the project.

## Budgeting

> The _Dev Team_ is allocated a budget, to be agreed with the Trustees. The scope of this budget is to cover costs incurred for development activities only, and not the cost of deploying any completed projects which would be arranged between the trustees and _Core Team_.

Reviewer 4: ```suggestion
The _Dev Team_ is allocated a budget, to be agreed with the Trustees. The scope of this budget is to cover costs incurred for development activities only, and not the cost of deploying any completed projects which would be arranged between the Trustees and _Core Team_.
```

Author: I disagree. One could argue that COBAL and python 2.3 are industry standard because they're probably used somewhere.

> For the purposes of budgetary accountability, the _Dev Team_ will select a Chair and Treasurer. (How the budget is managed is between the _Dev Team_ and the Trustees.) In practice, unlike the _Core Team_ budget which is fairly detailed (e.g. £x on Kickstart, £y on Competition, £z on other) and is approved in advance, _Dev Team_ purchasing will be defined at a higher level (maybe some “infrastructure” budget for a dev server and some dev tools but not £a for Project A, £b for Project B. Unlike the _Core Team_, this will not be a rolling budget, but a “pot” that gets topped up in each accounting period to allow for projects to be budgeted over multiple years.

Reviewer: > a “pot” that gets topped up in each accounting period

Having a pot-style budget is probably a bad idea. If the dev team don't use the budget, it should be re-allocated into the SR "pot". This also means if there's a project with a sizable development budget, that it is forced to go through the trustees.

Reviewer 6: @RealOrangeOne the problem with a non rolling pot is that funding for multi year projects cant be reserverd for the lifetime of the project. e.g. development of a new board that will take more than a year to complete and will have costs throughout the project.

Reviewer 5: Maybe project proposals should include a planned budget for the lifetime of the project. If the project proposal is accepted, funding for future years will be set aside in accordance with this budget. This would help avoid the situation where e.g. development of a new board has to pause because there is not enough money left in the current year's pot to fund manufacture.

Reviewer: I agree with @kierdavis that having something in the proposal about the expected budget is a good idea. 

I think we should defer to the trustees the actual implementation of how the budget works.

Author: > I think we should defer to the trustees the actual implementation of how the budget works.

Agreed. I think the core team should make a suggestion to them though.

Reviewer: Agreeed. I think what's written here is fine to cover that.

## Accountability

1. Much like _Core Team_, minutes and reports from the _Dev Team_, and its various sub-groups will be available to view by the general public. Sensitive data, such as budgeting, may be exempt from this.
2. The _Dev Team_ will submit regular progress reports to the Trustees.
3. As well as regular volunteers, the _Dev Team_ may contain trustees and members of the _Core Team_, however volunteers should bear in mind any potential conflicts of interest that may arise as a result of this.

> 4. Members of the _Dev Team_ should abstain from financial decisions where they are directly involved with the _project group_ of concern. In an event when all members of the _Dev Team_ have to abstain, the decision should be passed to the _Core Team_ or Trustees.

Reviewer: In the event the whole dev team abstains, the core team should be the next point of call.

Reviewer: Members of the dev team should abstain from any organisational votes for projects they are heavily involved in. If they are only involved in a support / review role, they probably don't need to

Author: We wrote `_Core Team_ or Trustees` as there are possible situations where the core team also should have to abstain. In this case the core team / trustees are simply acting as an external board because the dev team was unable to make a decision without a conflict of interest.

Reviewer: I agree, although we should make said calling order explicit. That the core team is contacted, _then_ the trustees, if the entire core team also abstain.

> 4. Members of the _Dev Team_ should abstain from financial decisions where they are directly involved with the _project group_ of concern. In an event when all members of the _Dev Team_ have to abstain, the decision should be passed to the _Core Team_ or Trustees.

Author: ```suggestion
4. Members of the _Dev Team_ should abstain from financial decisions where they are directly involved with the _project group_ of concern. In an event when all members of the _Dev Team_ have to abstain, the decision should be passed to the _Core Team_ and then Trustees.
```

5. The _Dev Team_ is responsible for the oversight of the various project groups within Student Robotics. The _project groups_ should report their progress regularly to the Dev Team.

## Relationship to the Core Team

- The _Dev Team_ is independent from the _Core Team_, but will work closely with them in certain situations.

> - The _Core Team_ will also communicate priorities for projects to be completed by the next competition cycle. In the event of a Kit Deployment, the _Core Team_ is responsible for budgeting, manufacture, deployment and testing.

Reviewer: Whilst I agree budgeting is the responsibility of the core team, i'm not sure I agree the rest are. In my experience, those designing and building a product (whether digital or physical) are best placed to manufacture and test it. The core team however is responsible for ensuring said testing happens.

If manufacturing requires outside equipment or parts, then ensuring they are ordered becomes the responsibility of the core team.

Author: The Core Team is responsible for the testing of the manufactured equipment. It is likely that they would ask they designers to do this testing, but we shouldn't make this too restrictive.

Reviewer: We should be explicit that the core team are responsible for ensuring it happens, with the help of the project group. The core team should not be the ones doing said testing etc

Author: I'm unsure exactly what you want. Please make a suggestion for how it should be worded?

Reviewer: ```suggestion
- The _Core Team_ will also communicate priorities for projects to be completed by the next competition cycle. In the event of a Kit Deployment, the _Core Team_ is responsible for ensuring budgeting, manufacture, deployment and testing is completed. The _project group_ will likely be the ones completing these tasks.
```

## Getting Things Done

For new projects, there is a process to follow to ensure they receive sufficient review and testing before they are implemented. This is shown in the examples below.

> For each _project group_ it is recommended that a “Project” is opened and actively maintained with high-level progress updates in the `srobo` GitHub - even the “nice to haves” - if there are good ideas being worked on which get parked here, at least there will be some documentation to pick up on.

Reviewer: ```suggestion
For each _project group_ it is recommended that a “Project” is opened and actively maintained with high-level progress updates in the `srobo` GitHub organisation - even the “nice to haves” - if there are good ideas being worked on which get parked here, at least there will be some documentation to pick up on.
```

Reviewer: Projects don't allow for much detail be provided for each point. I suspect that projects will fall into 2 categories. Either it's so simple that creating a project is too much work, or it's so large that a whole repository is required.

> Examples of what may be in scope for SR2020: Python 3, new brain board, new IDE.

Reviewer: This point isn't necessary in this proposal. That should be down to the development team, if they are operational in time for a 2020 shipment date.

Author: This serves simply as an example of what could be in scope.

I agree that it should be down development team to decide what is developed if this proposal is approved.

Reviewer: I think that's ok. Although it's the core team who decide if something actually ships for a given competition cycle.

> 

Reviewer: If the below are examples, it would be useful to have an 'examples' heading

Author: ```suggestion
## Examples
```

> Examples of what may be in scope for SR2020: Python 3, new brain board, new IDE.

Reviewer 3: Could we perhaps clarify this example by explicitly noting the relevant starting date here -- as this document will presumably form the basis of the definition of the Dev Team, the timings around this will become less clear over time.

Saying something like, "If they were operational around April 2019, the Dev Team might consider the following projects for SR2020" could convey much more clearly what this is saying.

Author: ```suggestion
If the Dev Team were operational in April 2019, the following are examples of what the Dev Team could consider to be in scope for SR2020: Python 3, new brain board, new IDE.
```

Reviewer: I still personally don't think this is necessary for a specification document. We have examples below. It adds nothing to the document

Author: I think it gives a useful example of a time scale. It was taken directly from @andybusse's proposal.

> ## Examples

Reviewer 3: I think it would be very useful to have an example of a task which originated with the Core Team. As the customer of the Dev Team's output, it's _very_ important that the Dev Team have a clear way to take input from them and work on the things which they find are needed.

Author: I will write one when I get a chance

### Project of significant impact: New brain board

> - A (group of) volunteer(s) decide(s) they’d like to work on finding a new single board computer which can serve as the brain board, and port/upgrade software to work on it.

Reviewer: As an example, we can just guess as to how many people are interested. It's not important. 

```suggestion
- A group of volunteers decide they’d like to work on finding a new single board computer which can serve as the brain board, and port/upgrade software to work on it.
```

> - New / Updated ticket in srobo/dev-tasks.

Reviewer: I suggest this repo be the same one used for the dev team minutes, in which case `srobo/dev-team` might be a better name?

Reviewer 2: I note that Core Team minutes go into a core team minutes repo, whereas tasks go into a tasks repo if I am not mistaken. Why should this be different for the dev team?

Reviewer: That's not quite correct. The tasks repo contains tasks for the entirety of SR, rather than just to do with competition operations and core team specific tasks.

Tasks specific to the core team _do_ go in the core team minutes repo.

Author: There should definitely be a separation from the Core Team's responsibilities. It also prevents spamming other teams with updates that they don't care about.

Reviewer 3: If this tasks repo is separate from https://github.com/srobo/tasks/, how do we co-ordinate development tasks which the Core Team need doing for the current ongoing competition? Would they go into both? (At which point we've got multiple copies of the same task) Or just into one? (At which point we've got multiple places which people need to look at in order to understand what needs doing)

- The volunteers develop a brief specification, including information like:
    - Functional requirements (e.g. compatible with the rest of the kit, can be powered from Power Board, runs our API, camera isn’t made slower, form factor)
    - Scope (e.g. excludes any changes to modular kit or API, independent of Python 3 project)
    - Deliverables (e.g. Board, Software, Documentation for students and volunteers)

>     - Testing requirements (e.g. is demonstrated to run a defined piece of software on a test robot without fault, with a certain response time for libkoki markers, etc.)

Reviewer: At work (which I think stems from the agile workflow), we refer to these as "Acceptance Criteria", that might be a better name for these?

>     - Testing requirements (e.g. is demonstrated to run a defined piece of software on a test robot without fault, with a certain response time for libkoki markers, etc.)

Author: ```suggestion
    - Acceptance Criteria (e.g. is demonstrated to run a defined piece of software on a test robot without fault, with a certain response time for libkoki markers, etc.)
```

- The _Dev Team_ may also define some limits, such as:

>     - Cut-off date for deployment at SR2020

Reviewer: The core team should define the cut off date for deployment. The dev team may define a cut off for consideration for a given competition cycle, but this should be disputable by the core team.

Author: As stated before, the _Dev Team_ will consult the _Core Team_, but as an independent group, they are free to ignore them.

Reviewer 2: Also note that as the Trustees ultimately approve the budget, they act as checks and balances.

The Dev Team are free to ignore **this year's** Core Team, and should have that right as their responsibility extends beyond that of the Core Team and so the interests of the two may be fundamentally different.

However, if the Dev Team decides to ignore everyone and just develop lots of pink machine-learning unicorns, it is unlikely they will be approved budget for this endeavor. This ensures that the Dev Team maintains freedom in development while also ensuring they can't do things which are plain stupid.

Reviewer: I think my original wording is a little off. Yes the dev team can consult the core team for a cut off date for a given project, but as the core team have the final say as to whether something ships in a given competition cycle, they may disagree with the given cut-off date.

Author: The Core Team ultimately makes the decision on whether something is shipped as stated elsewhere. I don't see any other issues here.

>     - Cost per unit limit

Reviewer: I thought the dev team was only responsible for development budget, this seems to go against that? This number may be required to see how viable a project is, in which case it should be consulted with the core team.

Author: Again, they will consult with the core team.

Reviewer: If this is such that they _will_ consult rather than _may_, this should be explicitly defined with these.

Author: They will consult the Core Team only if they deem it to be within the Core Team's scope and interests.

>     - Evidence shown that more than one board has been considered

Reviewer: What do you mean by this?

Author: evidence that multiple different solutions to the problem have been considered.

Reviewer 2: > multiple different solutions to the problem have been considered.

What if the problem is `pythonic` and so there exists only one obvious solution?

Author: > What if the problem is `pythonic` and so there exists only one obvious solution?

## Situation A: This is known when the _Dev Team_ define the requirements

Providing evidence that more than more solution has been considered will not be included in the requirements as the _Dev Team_ are aware that this is not possible.

## Situation B: This is not known when the _Dev Team_ define the requirements

The _project group_ will inform the core team with justification and request for the requirement to be approved. The _Dev Team_ will then consider if the justification is sufficient and then remove the requirement if it is deemed so.

Reviewer 2: In theory Situation B should not be possible if the sole solution is sufficiently obvious, as its obviousness would preclude it being unknown.

Author: Thence the vindication should be contemplated as sufficiently satisfactory.

Reviewer: I think this is infact fine. The wording above suggests that the dev team _may_ define these limits. To me that is also worded such that the project groups can contest said limits, which accounts for both of the above solutions.

- The _Dev Team_ may consult the _Core Team_ or Trustees to assist them with requirements. For example, in the case of a Project that would have a high implementation cost, the _Dev Team_ would need to communicate the feasibility of it with the Trustees.

> - This specification is presented to and ratified by the _Dev Team_ (either on the _Dev Team_ mailing list or at a _Dev Team_ meeting)

Reviewer: As the mailing list exists in a public form, and the specifications will include a budget, it might be worthwhile being explicit that those which contain a budget be conducted privately.

Author: ```suggestion
- This specification is presented to and ratified by the _Dev Team_ (either on the _Dev Team_ mailing list or at a _Dev Team_ meeting). Any sensitive information, such as a budget, will need to be privately conducted by the _Dev Team_ and _project group_.
```

Reviewer: > Any sensitive information, such as a budget, will need to be privately conducted 

Information cannot be conducted. 

```suggestion
- This specification is presented to and ratified by the _Dev Team_ (either on the _Dev Team_ mailing list or at a _Dev Team_ meeting). Any sensitive information, such as a budget, will need to be privately discussed by the _Dev Team_ and _project group_.
```

- Development work happens.
- If changes to the specification are needed, these can be discussed with the _Dev Team_.
- Development reaches a point where the project group is happy the work is done and is shippable.
- _Dev Team_ reviews the equipment against the agreed specification and decides whether to recommend it to the _Core Team_.

> - The process is now passed onto the _Core Team_ who are responsible for deciding when and how to deploy it (e.g for SR 2020).

Reviewer: The project group should define how _technically_ to deploy it. When, who and if are the remit of the core team.

>     - If it meets the needs of the competition, it is the _Core Team_’s responsibility to ensure new kit is purchased, tested and deployed (although it’s likely to be the _project group_ who do this).

Reviewer: I suggest that testing be done at least by the project group on the finished product, to ensure it meets their original specification. They're likely to know best as to how the product works, it's potential implementation risks.

Author: The _Core Team_ will have to work with the _project group_ to implement the testing, however it is the reponsibility of the core team to perform the testing.

Reviewer: Again, the core team shouldn't be performing the testing. The core team however should ensure that the project group do the testing (or the project group suggest alternative people to do the testing)

    - If it doesn’t, the _Core Team_ must decide whether to proceed anyway at risk, or to continue with the existing kit and pause expansion.

> ### Project of minor impact: Addition to the API (e.g. making an “R.ruggeduinos[0].PWM()” )

Reviewer 3: Thoughts on this example section:
- who is responsible for the overall view of the API, ensuring that it remains consistent throughout?
- I'm concerned that the Core Team (who are responsible for making sure that the competition works) have no explicit _knowledge_ of the change until it's been published; is the expectation that they would need to monitor the Dev Team's minutes to track upcoming changes? (If so, that might be ok, but let's be explicit about it)

Reviewer: It was my interpretation that once the project was completed, it wouldn't actually be published. The dev team would contact the core team saying something is ready, and then the core team work with the project group to ensure it ships correctly

Author: @RealOrangeOne is right here.

Reviewer 3: Ok, I think it would be good to be explicit about this relationship.

As this is notionally a simpler thing, but still in scope as it impacts the competition cycle, the process can be simpler.

> - Developer wants to add the feature and contacts the appropriate _project group_ or creates one if it does not exist.

Reviewer: What benefit does contacting an existing group bring? If there's an existing group rewriting the related code, and the shipment dates would match, then perhaps. But I suspect that level of project merging should be handled by the dev team rather than the volunteers themselves.
eg:
> "I want to add PWM to the ruggeduino in time for the next competition"
>
> "That's great! There's a team working on rewriting the ruggeduino firmware, which looks like it's going to be ready for the next competition, you should chat with them, see if they need any help!"

Author: Hence appropriate.

Reviewer: The wording is still not quite right. 
```suggestion
- Developer wants to add the feature and is pointed in the direction of the appropriate _project group_ by the _Dev Team, or creates one if it does not exist.
```

> - Developer wants to add the feature and is pointed in the direction of the appropriate _project group_ by the _Dev Team, or creates one if it does not exist.

Reviewer 3: nit
```suggestion
- Developer wants to add the feature and is pointed in the direction of the appropriate _project group_ by the _Dev Team_, or creates one if it does not exist.
```

- The _project group_ determines that there is no need for a specification due to minor impact.
    - A change of minor impact does not have significant cost or change to the working of the competition.
    - The _Dev Team_ may determine that a project is of minor impact and require a specification.

> - Developer updates/opens a new ticket in the appropriate GitHub Project.

Reviewer: In this case, should issues be created for the relevant repos (the brain board software and the ruggeduino firmware)?

Author: This isn't worded particularly well, GitHub allows issues on a repo to be added to a Project also.

Author: ```suggestion
- Developer updates/opens a new ticket on the appropriate GitHub Project, adding to the relevant repo also if one exists.
```

- The _project group_ will define testing and documentation requirements.
- Developer submits a pull request. 
- The _project group_ checks it meets the requirements and approves or rejects.

> - The _Dev Team_ will review the new version and decides whether to recommend it to the _Core Team_.

Reviewer 3: I'm concerned by the ordering here -- if at this point the Dev Team decide not to proposed the change to the Core Team, does that mean that the change has to be backed out of the repo (I'm not clear at what point we merge the changes to the repo)?
What happens if the Core Team for some reason don't want the change? (I'd hope this was rare)

Author: > I'm concerned by the ordering here -- if at this point the Dev Team decide not to proposed the change to the Core Team, does that mean that the change has to be backed out of the repo 

The Dev Team would work with the project group to make the appropriate changes for it to be shippable. In some cases this would mean reverting things. For example if the Project Group decides that the kit really badly needs support for flying pink unicorns, but also breaks vision.

>(I'm not clear at what point we merge the changes to the repo)?

The Project Group merges it when they approve. The Dev Team recommends a specific revision, i.e commit hash or release number to the Core Team

> What happens if the Core Team for some reason don't want the change? (I'd hope this was rare)

I would hope that the working together nature of the Dev Team would prevent this. If it did occur though, the Core Team should stick to the latest stable version that they are happy with.

- Software may be continuously developed throughout the competition, but it only deployed at the discretion of the _Core Team_. The _Core Team_ may choose to not upgrade software or equipment if it is not deemed beneficial.

> - When deploying, the _Core Team_ will update the kit to use the latest stable version of the software.

Reviewer: It's highly unlikely that the core team be the ones updating the kit. Once the core team give the green light, it'll be the reponsibility of the dev team / project group to ensure everything is in a shippable state, the core team to allocate someone to deploy it (this may be someone in the dev team / project group), and then it's shipped :shipit:

Author: In my opinion the _Dev Team_ should be constantly working on things and mark versions as stable. The Core Team will then deploy that version during prep for Kickstart etc.

Reviewer 2: This makes it unrealistic for, say, one very good software developer to update the kit. Let's say Steve the Software Developer writes some profoundly good firmware from the camera board as the only member of the project group. He must then deploy all the instances of this software, which if we have a lot of camera boards is an absurd quantity of work for one person.

In general, this appears to serve to massively discourage small group development, because that small group will then be incapable of deploying whatever they've developed, even though the fruits of their labour are perfectly good code / hardware.

Reviewer: That's true, perhaps the project group should work with the core team to ensure shipment happens correctly? Perhaps also by suggesting those who may be able to assist. 

In a realistic world, if you're developing a project, you have to be in a position to at least support it until the end, although obviously not solely. 

Related to https://github.com/srobo/dev-team-proposal/pull/1#discussion_r261861887

Author: In that instance the Core Team would ask for the project group to help them.

Nowhere does it say that the project groups cannot work directly with the core team if deemed necessary

Reviewer: > In that instance the Core Team would ask for the project group to help them.

We should define that in this example, as it's an important distinction.

Author: Suggested wording?

Reviewer: ```suggestion
- When deploying, the _Core Team_ will ensure the kit is updated with the newly created software, with help from the relevant _project group_.
```
